pandas/core/series.py

```diff
@@ -1105,11 +1116,17 @@ def _set_value(self, label, value, takeable: bool = False):
         takeable : interpret the index as indexers, default False
         """
         try:
-            if takeable:
+            dtype, _ = infer_dtype_from_scalar(value, pandas_dtype=True)
+            if takeable and is_dtype_equal(self.dtype, dtype):
                 self._values[label] = value
+            elif takeable:
+                self.iloc[label] = value
             else:
                 loc = self.index.get_loc(label)
                 validate_numeric_casting(self.dtype, value)
+                if not is_dtype_equal(self.dtype, dtype) and is_numeric_dtype(dtype):
```

Comment:
can you integrate this with the higher level if/else, these nested if/else are harder to read, even if it means duplicating some code

Reply:
Done